src/cdk/scrolling/virtual-scroll-viewport.ts

```diff
@@ -284,6 +284,9 @@ export class CdkVirtualScrollViewport extends CdkScrollable implements OnInit, O
 
   /** Sets the currently rendered range of indices. */
   setRenderedRange(range: ListRange) {
+    // The range should never be greater than the data.
+    range.end = Math.min(range.end, this.getDataLength());
```

Comment:
To keep it consistent, you could do the same for `range.start`. I think it would be good to not assume that the virtual scroll always starts at `0`.

```
range.start = Math.max(range.start, 0)
```

Reply:
Relates to https://github.com/angular/components/issues/23578